lib/matplotlib/category.py

```diff
@@ -23,6 +23,14 @@
 _log = logging.getLogger(__name__)
 
 
+def _check_axis(axis):
+    if axis is None:
+        raise units.ConversionError(
+            "Categorical data does not support unit conversion that is not "
+            "attached to an x or y Axis."
+        )
```

Comment:
This means no categorical color in this PR, because I didn't dive into the error it gave me... you raise a good point though that this PR should as a minimum not prevent categorical colorbars in the future. Let me take a look again at some point later this week, and I'll ping you with what I find.

Reply:
It was
```python
RuntimeError: <matplotlib.category.StrCategoryConverter object at 0x126bcd610> failed when trying to return the default units for this image. This may be because support has not been implemented for `axis=None` in the default_units() method.
```

because `StrCategoryConverter.default_units()` actually uses the `axis` argument (unlike all our other conversion interfaces.

I think this is fixable though, not sure why I just put those errors in... Don't have time now, but will chase categorical support next time I come back to this.

Reply:
Actually I fixed it 🎉 . Definitely don't have time to push the fix now, but glad it works!
```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
plt.imshow([["a", "b"], ["c", "d"]])
plt.colorbar()
plt.show()
```
![Figure_1](https://github.com/matplotlib/matplotlib/assets/6197628/df40809b-1ee7-42c1-9e39-672af69d3e8a)

Reply:
Right, this is in a good enough state (== tests passing) now if you want to branch off and play around with the colorbar for categoricals. I'm not intending to change anything there myself in this PR, but would be happy to merge something into this branch/PR if it's not too complicated.